pkg/k8s/graceful_shutdown.go

```diff
@@ -0,0 +1,71 @@
+package k8s
+
+import (
+	"context"
+	"time"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+)
+
+var log = logf.Log
+
+// StartStopSignalHandler starts gorutine which is waiting for
+// termination signal and returns chan for indication when operator
+// can really stop.
+func StartStopSignalHandler(client client.Client, namespaces []string) <-chan struct{} {
+	stopCH := make(chan struct{})
+	go handleStopSignal(client, namespaces, stopCH)
+	return stopCH
+}
+
+func handleStopSignal(client client.Client, namespaces []string, stopCH chan struct{}) {
+	<-signals.SetupSignalHandler()
+	stop(client, namespaces)
+	close(stopCH)
+}
+
+// Stop is used to understand, when we need to stop operator(usially SIGTERM)
+// to start cleanup process and delete required pxc clusters in current(operator)
+// namespace. See K8SPXC-529
+func stop(cl client.Client, namespaces []string) {
+	log.Info("Got stop signal, starting to list clusters")
+
+	for {
+		time.Sleep(5 * time.Second)
+		clustersAreReadyForDelete := true
+
+		for _, ns := range namespaces {
+
+			clusterList := &api.PerconaXtraDBClusterList{}
+
+			err := cl.List(context.TODO(), clusterList, &client.ListOptions{
+				Namespace: ns,
+			})
+			if err != nil {
+				log.Error(err, "list clusters in namespace", "namespace", ns)
+				continue
```

Comment:
Imagine we have namespace1,  namespace2 and namespace3.
For namespace 1 and 2 `List` operation ends with error. But for namespace3 it ends successfully.
Then namespace1 and 2 just ignored.

Reply:
what is the expected behavior?
Should we lock deleting if we cannot list clusters in namespace1 and namespace2?

Reply:
7559c1ef